gr-blocks/lib/repeat_impl.cc

```diff
@@ -24,50 +25,142 @@ repeat::sptr repeat::make(size_t itemsize, int interp)
-    set_msg_handler(pmt::mp("interpolation"),
-                    [this](pmt::pmt_t msg) { this->msg_set_interpolation(msg); });
+    if (interp < 1) {
+        throw std::invalid_argument(fmt::format(
+            FMT_STRING(
+                "Trying to set interpolation to {}, but must be a positive integer."),
+            interp));
+    }
+    message_port_register_in(c_msg_port);
+    set_msg_handler(c_msg_port,
+                    [this](const pmt::pmt_t& msg) { this->msg_set_interpolation(msg); });
 }
 
-void repeat_impl::msg_set_interpolation(pmt::pmt_t msg)
+void repeat_impl::msg_set_interpolation(const pmt::pmt_t& msg)
 {
-    // Dynamization by Kevin McQuiggin:
-    d_interp = pmt::to_long(pmt::cdr(msg));
-    sync_interpolator::set_interpolation(d_interp);
+    if (!pmt::eq(c_msg_port, pmt::car(msg))) {
+        d_logger->error("Message with unknown key {}", pmt::car(msg));
+    }
+    auto new_interp = pmt::to_long(pmt::cdr(msg));
+    if (new_interp < 0) {
+        d_logger->error(
+            "Trying to set interpolation to {}, but must be a positive integer.",
+            new_interp);
+        return;
+    }
+    d_logger->trace("Setting new interpolation {}, replacing {}. Items left at old "
+                    "interpolation: {}.",
+                    new_interp,
+                    d_interp,
+                    d_left_to_copy);
+    d_interp = new_interp;
 }
 void repeat_impl::set_interpolation(int interp)
 {
     // This ensures that interpolation is only changed between calls to work
     // (and not in the middle of an ongoing work)
-    _post(pmt::mp("interpolation"),                                   /* port */
-          pmt::cons(pmt::mp("interpolation"), pmt::from_long(interp)) /* pair */
+    if (interp < 1) {
+        d_logger->error(
+            "Trying to set interpolation to {}, but must be a positive integer.", interp);
+        throw std::invalid_argument(fmt::format(
+            FMT_STRING(
+                "Trying to set interpolation to {}, but must be a positive integer."),
+            interp));
+    }
+
+    _post(c_msg_port,                                   /* port */
+          pmt::cons(c_msg_port, pmt::from_long(interp)) /* pair */
     );
 }
 
-int repeat_impl::work(int noutput_items,
-                      gr_vector_const_void_star& input_items,
-                      gr_vector_void_star& output_items)
+void repeat_impl::forecast(int noutput_items, gr_vector_int& required)
 {
+    if (static_cast<size_t>(noutput_items) <= d_left_to_copy) {
+        // can fill request from memory
+        required[0] = 0;
+        return;
+    }
+
+
+    // making sure we get enough input
+    size_t requested_round_up = (static_cast<size_t>(noutput_items) -
+                                 d_left_to_copy) // amount we haven't "prepared"
+                                + d_interp - 1;  // round up to multiple of d_interp
+
+    // only one possible input
+    required[0] = requested_round_up / d_interp;
```

Comment:
Example: nout = 10, left to copy = 3, interp = 4
10 - 3 + 4 - 1 = 10
10 / 4 = 2
Should the answer be 1? Were the parens meant to do something here?